<?xml version="1.0" ?>
<sdf version="1.4">
  <world name="default">
    <physics type="ode">
      <gravity>0 0 0</gravity>
      <ode>
        <solver>
          <type>quick</type>
          <iters>1000</iters>
          <precon_iters>0</precon_iters>
          <sor>1.000000</sor>
        </solver>
        <constraints>
          <cfm>0.000000</cfm>
          <erp>0.200000</erp>
          <contact_max_correcting_vel>100.000000</contact_max_correcting_vel>
          <contact_surface_layer>0.01000</contact_surface_layer>
        </constraints>
      </ode>
      <bullet>
        <solver>
          <type>sequential_impulse</type>
          <iters>1000</iters>
          <sor>1.000000</sor>
        </solver>
        <constraints>
          <cfm>0.000000</cfm>
          <erp>0.200000</erp>
          <split_impulse>true</split_impulse>
          <split_impulse_penetration_threshold>-0.01</split_impulse_penetration_threshold>
          <contact_surface_layer>0.00000</contact_surface_layer>
        </constraints>
      </bullet>
      <real_time_update_rate>0.000000</real_time_update_rate>
      <max_step_size>0.001</max_step_size>
    </physics>
    <include>
      <uri>model://sun</uri>
    </include>

    <!-- to demonstrate force torque, we'll construct a model with
    two bodies stacked vertically, with a x-revolute joint connecting
    them.  The joint has 90 degree limit.  We'll test force
    torque readings and characterize them. -->
    <model name="model_1">
      <link name="link_1">
        <pose>0 0 0.5 0 0 0</pose>
        <inertial>
          <pose>0 0 0.5 0 0 0</pose>
          <inertia>
            <ixx>0.866666</ixx>
            <ixy>0.000000</ixy>
            <ixz>0.000000</ixz>
            <iyy>0.841666</iyy>
            <iyz>0.000000</iyz>
            <izz>0.041666</izz>
          </inertia>
          <mass>10.000000</mass>
        </inertial>
        <visual name="visual_box">
          <pose>0 0 0.5 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 0.2 1.0</size>
            </box>
          </geometry>
        </visual>
        <collision name="collision_box">
          <pose>0 0 0.5 0 0 0</pose>
          <max_contacts>250</max_contacts>
          <geometry>
            <box>
              <size>0.1 0.2 1.0</size>
            </box>
          </geometry>
        </collision>
      </link>

      <link name="link_2">
        <!-- 1cm gap between link_1 and link_2 -->
        <pose>0 0 1.51 0 0 0</pose>
        <inertial>
          <pose>0 0 0.5 0 0 0</pose>
          <inertia>
            <ixx>0.866666</ixx>
            <ixy>0.000000</ixy>
            <ixz>0.000000</ixz>
            <iyy>0.841666</iyy>
            <iyz>0.000000</iyz>
            <izz>0.041666</izz>
          </inertia>
          <mass>10.000000</mass>
        </inertial>
        <visual name="visual_box">
          <pose>0 0 0.5 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 0.2 1.0</size>
            </box>
          </geometry>
        </visual>
        <collision name="collision_box">
          <pose>0 0 0.5 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 0.2 1.0</size>
            </box>
          </geometry>
        </collision>
      </link>

      <link name="link_3">
        <pose>0 0 2.62 0 0 0</pose>
        <inertial>
          <pose>0 0.5 0 0 0 0</pose>
          <inertia>
            <ixx>0.866666</ixx>
            <ixy>0.000000</ixy>
            <ixz>0.000000</ixz>
            <iyy>0.041666</iyy>
            <iyz>0.000000</iyz>
            <izz>0.841666</izz>
          </inertia>
          <mass>10.000000</mass>
        </inertial>
        <visual name="visual_box">
          <pose>0 0.5 0 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 1.0 0.2</size>
            </box>
          </geometry>
        </visual>
        <collision name="collision_box">
          <pose>0 0 0.5 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 1.0 0.2</size>
            </box>
          </geometry>
        </collision>
      </link>

      <link name="link_4">
        <pose>0 1.01 2.62 0 0 0</pose>
        <inertial>
          <pose>0 0.5 0 0 0 0</pose>
          <inertia>
            <ixx>0.866666</ixx>
            <ixy>0.000000</ixy>
            <ixz>0.000000</ixz>
            <iyy>0.041666</iyy>
            <iyz>0.000000</iyz>
            <izz>0.841666</izz>
          </inertia>
          <mass>10.000000</mass>
        </inertial>
        <visual name="visual_box">
          <pose>0 0.5 0 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 1.0 0.2</size>
            </box>
          </geometry>
        </visual>
        <collision name="collision_box">
          <pose>0 0 0.5 0 0 0</pose>
          <geometry>
            <box>
              <size>0.1 1.0 0.2</size>
            </box>
          </geometry>
        </collision>
      </link>

      <joint name="joint_01" type="revolute">
        <parent>world</parent>
        <child>link_1</child>
        <!-- joint at origin of link_1 inertial frame -->
        <!-- moement arm from link_1 inertial frame to joint_01 is 0m -->
        <pose>0 0 0.0 0 0 0</pose>
        <axis>
          <limit>
            <lower>-1.57079</lower>
            <upper>1.57079</upper>
            <effort>1000.000000</effort>
            <velocity>-1.000000</velocity>
            <!-- below is only for simbody. For ode, it's under
                 <physics><ode><limit><erp/><cfm/></limit></ode></physics>
                 TODO: Consider unifying the two ways of specifying
                       joint limit properties to here.
            -->
            <stiffness>1e8</stiffness>
            <dissipation>1.0</dissipation>
          </limit>
          <dynamics>
            <damping>0.000000</damping>
            <friction>0.000000</friction>
          </dynamics>
          <xyz>1.000000 0.000000 0.000000</xyz>
        </axis>
        <sensor name="force_torque" type="force_torque">
          <always_on>true</always_on>
          <visualize>true</visualize>
          <update_rate>30</update_rate>
        </sensor>
        <physics>
          <provide_feedback>true</provide_feedback>
          <implicit_spring_damper>true</implicit_spring_damper>
          <ode>
            <!-- for constrained directions -->
            <cfm>0.0</cfm>
            <erp>0.2</erp>
            <!-- bounce is for limits, consider moving under <limit> -->
            <bounce>0.0</bounce>
            <limit>
              <!-- for limits -->
              <cfm>0.0</cfm>
              <erp>0.2</erp>
            </limit>
          </ode>
        </physics>
      </joint>

      <joint name="joint_12" type="revolute">
        <parent>link_1</parent>
        <child>link_2</child>
        <!-- joint_1 at origin of link_2 link frame -->
        <!-- moement arm from link_2 inertial frame to joint_01 is 2m -->
        <!-- moement arm from link_2 inertial frame to joint_12 is 0.5m -->
        <pose>0 0 0 0 0 0</pose>
        <axis>
          <limit>
            <lower>-4.0</lower>
            <upper>4.0</upper>
            <effort>10000.000000</effort>
            <velocity>-1.000000</velocity>
            <!-- below is only for simbody. For ode, it's under
                 <physics><ode><limit><erp/><cfm/></limit></ode></physics>
                 TODO: Consider unifying the two ways of specifying
                       joint limit properties to here.
            -->
            <stiffness>1e8</stiffness>
            <dissipation>1.0</dissipation>
          </limit>
          <dynamics>
            <damping>0.000000</damping>
            <friction>0.000000</friction>
          </dynamics>
          <xyz>0.000000 0.000000 1.000000</xyz>
        </axis>
        <sensor name="force_torque" type="force_torque">
          <always_on>true</always_on>
          <visualize>true</visualize>
          <update_rate>30</update_rate>
        </sensor>
        <physics>
          <provide_feedback>true</provide_feedback>
          <implicit_spring_damper>true</implicit_spring_damper>
          <ode>
            <!-- for constrained directions -->
            <cfm>0.0</cfm>
            <erp>0.2</erp>
            <!-- bounce is for limits, consider moving under <limit> -->
            <bounce>0.0</bounce>
            <limit>
              <!-- for limits -->
              <cfm>0.0</cfm>
              <erp>0.2</erp>
            </limit>
          </ode>
        </physics>
      </joint>

      <joint name="joint_23" type="revolute">
        <parent>link_2</parent>
        <child>link_3</child>
        <!-- joint_1 at origin of link_2 link frame -->
        <!-- moement arm from link_2 inertial frame to joint_01 is 2m -->
        <!-- moement arm from link_2 inertial frame to joint_12 is 0.5m -->
        <pose>0 0 0 0 0 0</pose>
        <axis>
          <limit>
            <lower>-1.0</lower>
            <upper>1.0</upper>
            <effort>10000.000000</effort>
            <velocity>-1.000000</velocity>
            <!-- below is only for simbody. For ode, it's under
                 <physics><ode><limit><erp/><cfm/></limit></ode></physics>
                 TODO: Consider unifying the two ways of specifying
                       joint limit properties to here.
            -->
            <stiffness>1e8</stiffness>
            <dissipation>1.0</dissipation>
          </limit>
          <dynamics>
            <damping>0.000000</damping>
            <friction>0.000000</friction>
          </dynamics>
          <xyz>0.000000 0.000000 1.000000</xyz>
        </axis>
        <sensor name="force_torque" type="force_torque">
          <always_on>true</always_on>
          <visualize>true</visualize>
          <update_rate>30</update_rate>
        </sensor>
        <physics>
          <provide_feedback>true</provide_feedback>
          <implicit_spring_damper>true</implicit_spring_damper>
          <ode>
            <!-- for constrained directions -->
            <cfm>0.0</cfm>
            <erp>0.2</erp>
            <!-- bounce is for limits, consider moving under <limit> -->
            <bounce>0.0</bounce>
            <limit>
              <!-- for limits -->
              <cfm>0.0</cfm>
              <erp>0.2</erp>
            </limit>
          </ode>
        </physics>
      </joint>

      <joint name="joint_34" type="revolute">
        <parent>link_3</parent>
        <child>link_4</child>
        <!-- joint_1 at origin of link_2 link frame -->
        <!-- moement arm from link_2 inertial frame to joint_01 is 2m -->
        <!-- moement arm from link_2 inertial frame to joint_12 is 0.5m -->
        <pose>0 0 0 0 0 0</pose>
        <axis>
          <limit>
            <lower>-1.0</lower>
            <upper>1.0</upper>
            <effort>10000.000000</effort>
            <velocity>-1.000000</velocity>
            <!-- below is only for simbody. For ode, it's under
                 <physics><ode><limit><erp/><cfm/></limit></ode></physics>
                 TODO: Consider unifying the two ways of specifying
                       joint limit properties to here.
            -->
            <stiffness>1e8</stiffness>
            <dissipation>1.0</dissipation>
          </limit>
          <dynamics>
            <damping>0.000000</damping>
            <friction>0.000001</friction>
          </dynamics>
          <xyz>0.000000 0.000000 1.000000</xyz>
        </axis>
        <sensor name="force_torque" type="force_torque">
          <always_on>true</always_on>
          <visualize>true</visualize>
          <update_rate>30</update_rate>
        </sensor>
        <physics>
          <provide_feedback>true</provide_feedback>
          <implicit_spring_damper>true</implicit_spring_damper>
          <ode>
            <!-- for constrained directions -->
            <cfm>0.0</cfm>
            <erp>0.2</erp>
            <!-- bounce is for limits, consider moving under <limit> -->
            <bounce>0.0</bounce>
            <limit>
              <!-- for limits -->
              <cfm>0.0</cfm>
              <erp>0.2</erp>
            </limit>
          </ode>
        </physics>
      </joint>
    </model>
  </world>
</sdf>
